lib/luacrypto.c

```diff
@@ -0,0 +1,438 @@
+	struct scatterlist sg[2];
+
+	struct crypto_aead *tfm = luacrypto_check(L, 1);
+	const char *iv = luaL_checklstring(L, 2, &iv_len);
+	const char *plaintext = luaL_checklstring(L, 3, &pt_len);
+	aad_len = 0;  // aad is optional
+	const char *aad = (!lua_isnoneornil(L, 4)) ? luaL_checklstring(L, 4, &aad_len) : NULL;
+
+	int ret;
+
+	authsize = crypto_aead_authsize(tfm);
+	iv_len_expected = crypto_aead_ivsize(tfm);
+	if (iv_len != iv_len_expected)
+	return luaL_error(
+		L, "incorrect nonce length: expected %zu, got %zu",
+		iv_len_expected, iv_len
+	);
+	ct_len = pt_len + authsize;
+
+	iv_buf = kmemdup(iv, iv_len, gfp);
+	pt_buf = kmalloc(ct_len, gfp);
+	if (aad_len > 0) {
+		aad_buf = kmemdup(aad, aad_len, gfp);
+	} else {
+		aad_buf = NULL;
+	}
+	if (!iv_buf || !pt_buf || (aad_len > 0 && !aad_buf)) {
+		kfree(aad_buf);
+		kfree(iv_buf);
+		kfree(pt_buf);
+		return luaL_error(L, "failed to allocate buffers");
+	}
+	memcpy(pt_buf, plaintext, pt_len);
+
+	req = aead_request_alloc(tfm, gfp);
+	if (!req) {
+		kfree(aad_buf);
+		kfree(iv_buf);
+		kfree(pt_buf);
+		return luaL_error(L, "failed to allocate aead request");
+	}
+
+	if (aad_len > 0) {
+		sg_init_table(sg, 2);
+		sg_set_buf(&sg[0], aad_buf, aad_len);
+		sg_set_buf(&sg[1], pt_buf, ct_len);
+	} else {
+		sg_init_table(sg, 1);
+		sg_set_buf(&sg[0], pt_buf, ct_len);
+	}
+
+	aead_request_set_ad(req, aad_len);
+	aead_request_set_callback(req, 0, NULL, NULL);
+	aead_request_set_crypt(req, sg, sg, pt_len, iv_buf);
+	ret = crypto_aead_encrypt(req);
+
+	aead_request_free(req);
+	kfree(aad_buf);
+	kfree(iv_buf);
+
+	if (ret) {
+		kfree(pt_buf);
+		return luaL_error(L, "encryption failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)pt_buf, ct_len);
+	lua_pushinteger(L, authsize);
+	kfree(pt_buf);
+	return 2;
+}
+
+/*
+* crypto:decrypt(nonce, ciphertext, aad)
+*
+* Decrypts the ciphertext using the provided nonce and additional authenticated data (AAD).
+* The ciphertext must include the appended authentication tag.
+* On success, returns the plaintext.
+*/
+static int luacrypto_decrypt(lua_State *L)
+{
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));
+	size_t iv_len, ct_len, aad_len, pt_len, authsize, iv_len_expected;
+	u8 *iv_buf, *ct_buf, *aad_buf;
+	struct aead_request *req;
+	struct scatterlist sg[2];
+
+	struct crypto_aead *tfm = luacrypto_check(L, 1);
+	const char *iv = luaL_checklstring(L, 2, &iv_len);
+	const char *ciphertext = luaL_checklstring(L, 3, &ct_len);
+	aad_len = 0;
+	const char *aad = (!lua_isnoneornil(L, 4)) ? luaL_checklstring(L, 4, &aad_len) : NULL;
+	int ret;
+
+
+	iv_len_expected = crypto_aead_ivsize(tfm);
+	if (iv_len != iv_len_expected)
+		return luaL_error(
+			L, "incorrect nonce length: expected %zu, got %zu",
+			iv_len_expected, iv_len
+		);
+
+	authsize = crypto_aead_authsize(tfm);
+	if (ct_len < authsize)
+		return luaL_error(L, "ciphertext too short (must include tag)");
+
+	pt_len = ct_len - authsize;
+
+	iv_buf = kmemdup(iv, iv_len, gfp);
```

Comment:
you need to check for NULL, right?

Reply:
I do it for all buffers on l. 315. If you prefer, I can do it right after each one.

Reply:
thing is, if this allocation fail, you're gonna allocate the next one and you're going to kree(NULL), you're not supposed to do that..